modules/ardrivo/src/Servo.cxx

```diff
@@ -0,0 +1,88 @@
+/*
+ *  Servo.cxx
+ *  Copyright 2020 ItJustWorksTM
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ */
+
+#include <Arduino.h>
+#include "BoardDataDef.hxx"
+#include "Servo.h"
+
+#define SERVO_MIN() (MIN_PULSE_WIDTH - this->min * 4)
+#define SERVO_MAX() (MAX_PULSE_WIDTH - this->max * 4)
+
+#define TRIM_DURATION 2
+
+uint32_t servoCount = 0;
+
+Servo::Servo() {
+    if (servoCount < board_data->servo_value.max_size()) {
+        this->servoIndex = servoCount++;
+        board_data->servo_value[this->servoIndex].angle = 90;
+    } else
+        this->servoIndex = INVALID_SERVO; // too many servos
+}
+
+uint8_t Servo::attach(int pin) { return this->attach(pin, MIN_PULSE_WIDTH, MAX_PULSE_WIDTH); }
+
+uint8_t Servo::attach(int pin, int min, int max) {
+    if (this->servoIndex < board_data->servo_value.max_size()) {
+        board_data->servo_value[this->servoIndex].Pin.nbr = pin;
+        this->min = (MIN_PULSE_WIDTH - min) / 4;
+        this->max = (MAX_PULSE_WIDTH - max) / 4;
+        board_data->servo_value[this->servoIndex].Pin.isActive = true;
+    }
+    return this->servoIndex;
+}
+
+void Servo::detach() { board_data->servo_value[this->servoIndex].Pin.isActive = false; }
+
+void Servo::write(int value) {
+    if (value < 0) {
+        value = 0;
+    }
+    if (value > 180) {
+        value = 180;
+    }
+    board_data->servo_value[this->servoIndex].angle = value;
+}
+
+void Servo::writeMicroseconds(int value) {
+    if (this->servoIndex < board_data->servo_value.max_size()) {
+        if (value < SERVO_MIN())
+            value = SERVO_MIN();
+        else if (value > SERVO_MAX())
+            value = SERVO_MAX();
+
+        value = value - TRIM_DURATION;
+        value = map(value, MIN_PULSE_WIDTH, MAX_PULSE_WIDTH, 0, 180); // Convert microseconds to degree
+
+        this->write(value);
+    }
+}
+
+int Servo::read() { return board_data->servo_value[this->servoIndex].angle; }
+
+int Servo::readMicroseconds() {
+    unsigned int angle;
+    if (this->servoIndex != INVALID_SERVO)
+        angle = board_data->servo_value[this->servoIndex].angle;
+    else
+        angle = 0;
+
+    return map(angle, 0, 180, MIN_PULSE_WIDTH, MAX_PULSE_WIDTH); // Convert degree to microseconds
+}
+
+bool Servo::attached() { return board_data->servo_value[this->servoIndex].Pin.isActive; }
```

Comment:
I'd pretty much expect this function to only inspect the object's state and definitely not to go take a look at global state